### ISSUE TITLE
Saving/Loading parameters changes weight format and can lead to silent errors

### ISSUE THREAD

**Issue by patrickvonplaten** [google/flax#1261]
Provide as much information as possible. At least, this should include a description of your issue and steps to reproduce the problem. If possible also provide a summary of what steps or workarounds you have already tried.

### Problem you have encountered:

If one saves & reloads a `state` as follows:

```python
  model_file = os.path.join(temp_dir, "flax_model")
  with open(model_file, "wb") as f:
      model_bytes = to_bytes(state)
      f.write(model_bytes)

  with open(model_file, "rb") as state_f:
      state = from_bytes(AwesomeModel, state_f.read())
```

The `state` previously contained weights of type `jaxlib.xla_extension.DeviceArray`, but is then changed to simply numpy arrays: `np.ndarray`. 

This can lead to silent bugs when, *e.g.*, an embedding layer `flax.linen.Embed` is used since it behaves differently depending on the format of its weights.

### What you expected to happen:

I would expect saved weights to be loaded in the exact same format. Or maybe am I using the wrong save/load functions?

### Steps to reproduce:
If you run the following code, you can see how saving/loading can lead to unexpected errors:

```python
#!/usr/bin/env python3
import flax.linen as nn
from flax.serialization import to_bytes, from_bytes
import numpy as np
import jax.numpy as jnp
from jax.random import PRNGKey
import tempfile
import os


class AwesomeModel(nn.Module):
    dtype: jnp.dtype = jnp.float32
    num_embeddings = 20
    embedding_size = 16

    def setup(self):
        self.word_emb = nn.Embed(self.num_embeddings, self.embedding_size, dtype=self.dtype)

    def __call__(self, input_ids):
        return self.word_emb(input_ids)


input_ids = jnp.broadcast_to(np.arange(10)[None, :], (4, 10))  # => shape [4, 10]

model = AwesomeModel()
state = model.init(PRNGKey(0), input_ids)

# No problem here!
embeddings = model.apply(state, input_ids)

# Now let's save & reload the weights
with tempfile.TemporaryDirectory() as temp_dir:
    model_file = os.path.join(temp_dir, "flax_model")
    with open(model_file, "wb") as f:
        model_bytes = to_bytes(state)
        f.write(model_bytes)

    with open(model_file, "rb") as state_f:
        state = from_bytes(AwesomeModel, state_f.read())

# ...and rerun the forward pass
embeddings = model.apply(state, input_ids) # Error - this is a bit unexpected
```

**Comment by patrickvonplaten** [comment]
This also seems to be linked to: https://github.com/google/flax/issues/1157

**Comment by marcvanzee** [comment]
Thanks for the nice repro!

This is indeed unexpected: we use indexing in the Embedding module, but that doesn't work well if the indexer and indexee are of different types. In your case the indexee is np.array and the indexer is jnp.array, which doesn't seem to work.

I'll look into a fix for this, but in the meantime to unblock you, you can just map your params to `jnp.array` after reading them from your file:

```
    with open(model_file, "rb") as state_f:
        state = from_bytes(AwesomeModel, state_f.read())
        state = jnp.tree_map(jnp.array, state)
```

**Comment by patrickvonplaten** [comment]
Thanks a lot for the pointer @marcvanzee ! 

I think the function is called `jax.tree_util.tree_map` instead of `jnp.tree_map` no?

**Comment by marcvanzee** [comment]
Yes sorry about that, `jax.tree_map` also works, I've updated the example. By the way, @jheek fixed this issue in #1262.
